mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulPlayScreen.kt

```diff
@@ -1,34 +1,189 @@
 package ch.epfl.sdp.mobile.state
 
 import androidx.compose.foundation.layout.PaddingValues
-import androidx.compose.runtime.Composable
-import androidx.compose.runtime.remember
+import androidx.compose.runtime.*
 import androidx.compose.ui.Modifier
+import ch.epfl.sdp.mobile.application.authentication.AuthenticatedUser
+import ch.epfl.sdp.mobile.application.chess.ChessFacade
+import ch.epfl.sdp.mobile.application.chess.Match
+import ch.epfl.sdp.mobile.application.chess.engine.Color
+import ch.epfl.sdp.mobile.application.chess.engine.NextStep
+import ch.epfl.sdp.mobile.application.chess.notation.serialize
+import ch.epfl.sdp.mobile.state.Loadable.Companion.loaded
+import ch.epfl.sdp.mobile.state.Loadable.Companion.loading
 import ch.epfl.sdp.mobile.ui.play.PlayScreen
 import ch.epfl.sdp.mobile.ui.play.PlayScreenState
+import ch.epfl.sdp.mobile.ui.social.*
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
 
+/** An Loadable Union Type to differentiate between loaded and state */
+sealed interface Loadable<out T> {
+  object Loading : Loadable<Nothing>
+  data class Loaded<out T>(val value: T) : Loadable<T>
+
+  companion object {
+    fun loading(): Loadable<Nothing> = Loading
+    fun <T> loaded(value: T): Loadable<T> = Loaded(value)
+  }
+}
 /**
- * Player class that holds info about chess player
- * @param newGame Callable when new game button is actioned. Switches to pregame screen
+ * A map for the Loadable type
+ * @param f callback function for map
  */
-/* FIXME: add more useful args : user... */
-class Player(private val newGame: () -> Unit) : PlayScreenState {
-  /* TODO: override other attributes */
-  override val onNewGameClick = newGame
+private inline fun <A, B> Loadable<A>.map(f: (A) -> B): Loadable<B> =
+    when (this) {
+      is Loadable.Loaded -> loaded(f(value))
+      Loadable.Loading -> loading()
+    }
+
+/**
+ * Extract the value out of the Loadable type
+ * @param lazyBlock function which returns alternative value
+ */
+private inline fun <A> Loadable<A>.orElse(lazyBlock: () -> A): A =
+    when (this) {
+      is Loadable.Loaded -> value
+      Loadable.Loading -> lazyBlock()
+    }
+
+/**
+ * Gives higher order function to decide the MatchResult
+ * @param userColor color of the current user
+ */
+private inline fun NextStep.toMatchResult(): (userColor: Color) -> MatchResult =
+    when (this) {
+      is NextStep.Stalemate -> { _ -> Tie }
+      is NextStep.MovePiece -> { userColor -> if (this.turn == userColor) YourTurn else OtherTurn }
+      is NextStep.Checkmate -> { userColor ->
+            if (userColor == this.winner) {
+              Win(MatchResult.Reason.CHECKMATE)
+            } else {
+              Loss(MatchResult.Reason.CHECKMATE)
+            }
+          }
+    }
```

Comment:
Any reason for this function to be `inline` ? It doesn't seem to have lambda arguments or need reified generics.

Nice use of a function return type though !